[0.10.0] Crashed the server with a bad derivative query

* centos 6.7 
* influxdb-0.10.0-1.x86_64

Ran the following query from Grafana - crashed the influxdb server

```
SELECT (derivative(value, 1m) / 60.000) FROM collectd."default".interface_rx 
WHERE type = 'if_octets' AND type_instance = 'eth0' AND time > now() - 30m AND 
time < now() - 1m GROUP BY host
```

Log output:

```
[query] 2016/02/11 10:22:15 SELECT (derivative(value, 1m) / 60.000) FROM collectd."default".interface_rx WHERE type = 'if_octets' AND type_instance = 'eth0' AND time > now() - 30m AND time < now() - 1m GROUP BY host
panic: interface conversion: interface is float64, not map[string]interface {}

goroutine 233276 [running]:
github.com/influxdb/influxdb/tsdb.(*limitedRowWriter).processValues(0xc20b1afd40, 0xc20c796800, 0xad, 0x100, 0x0)
        /tmp/tmp.NZsW7kxYIE/src/github.com/influxdb/influxdb/tsdb/raw.go:524 +0x1173
github.com/influxdb/influxdb/tsdb.(*limitedRowWriter).Flush(0xc20b1afd40)
        /tmp/tmp.NZsW7kxYIE/src/github.com/influxdb/influxdb/tsdb/raw.go:446 +0xd6
github.com/influxdb/influxdb/tsdb.(*RawExecutor).execute(0xc20baba7b0, 0xc2093420c0, 0xc20805ff20)
        /tmp/tmp.NZsW7kxYIE/src/github.com/influxdb/influxdb/tsdb/raw.go:154 +0x17e5
created by github.com/influxdb/influxdb/tsdb.(*RawExecutor).Execute
        /tmp/tmp.NZsW7kxYIE/src/github.com/influxdb/influxdb/tsdb/raw.go:59 +0x64

```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -20,6 +20,8 @@
 - [#5375](https://github.com/influxdata/influxdb/pull/5375): Lint tsdb and tsdb/engine package @nuss-justin
 - [#5624](https://github.com/influxdata/influxdb/pull/5624): Fix golint issues in client v2 package @PSUDaemon
 - [#5510](https://github.com/influxdata/influxdb/pull/5510): Optimize ReducePercentile @bsideup
+- [#5557](https://github.com/influxdata/influxdb/issues/5630): Fixes panic when surrounding the select statement arguments in brackets
+- [#5628](https://github.com/influxdata/influxdb/issues/5628): Crashed the server with a bad derivative query
 
 ## v0.10.0 [2016-02-04]
 

--- a/influxql/select.go
+++ b/influxql/select.go
@@ -137,7 +137,8 @@ func buildFieldIterators(fields Fields, ic IteratorCreator, opt IteratorOptions)
 				continue
 			}
 
-			itr, err := buildExprIterator(f.Expr, ic, opt)
+			expr := Reduce(f.Expr, nil)
+			itr, err := buildExprIterator(expr, ic, opt)
 			if err != nil {
 				return err
 			}
@@ -163,7 +164,8 @@ func buildFieldIterators(fields Fields, ic IteratorCreator, opt IteratorOptions)
 				continue
 			}
 
-			itr, err := buildExprIterator(f.Expr, aitr, opt)
+			expr := Reduce(f.Expr, nil)
+			itr, err := buildExprIterator(expr, aitr, opt)
 			if err != nil {
 				return err
 			}
@@ -358,6 +360,8 @@ func buildExprIterator(expr Expr, ic IteratorCreator, opt IteratorOptions) (Iter
 			}
 			return buildTransformIterator(lhs, rhs, expr.Op, ic, opt)
 		}
+	case *ParenExpr:
+		return buildExprIterator(expr.Expr, ic, opt)
 	default:
 		panic(fmt.Sprintf("invalid expression type: %T", expr)) // FIXME
 	}

--- a/influxql/select_test.go
+++ b/influxql/select_test.go
@@ -1417,3 +1417,60 @@ func TestSelect_BinaryExpr_Integer(t *testing.T) {
 		}
 	}
 }
+
+// Ensure a SELECT (...) query can be executed.
+func TestSelect_ParenExpr(t *testing.T) {
+	var ic IteratorCreator
+	ic.CreateIteratorFn = func(opt influxql.IteratorOptions) (influxql.Iterator, error) {
+		if !reflect.DeepEqual(opt.Expr, MustParseExpr(`min(value)`)) {
+			t.Fatalf("unexpected expr: %s", spew.Sdump(opt.Expr))
+		}
+
+		return influxql.NewCallIterator(&FloatIterator{Points: []influxql.FloatPoint{
+			{Name: "cpu", Tags: ParseTags("region=west,host=A"), Time: 0 * Second, Value: 20},
+			{Name: "cpu", Tags: ParseTags("region=west,host=B"), Time: 5 * Second, Value: 10},
+			{Name: "cpu", Tags: ParseTags("region=east,host=A"), Time: 9 * Second, Value: 19},
+			{Name: "cpu", Tags: ParseTags("region=east,host=A"), Time: 10 * Second, Value: 2},
+			{Name: "cpu", Tags: ParseTags("region=west,host=A"), Time: 11 * Second, Value: 3},
+			{Name: "cpu", Tags: ParseTags("region=west,host=A"), Time: 31 * Second, Value: 100},
+		}}, opt), nil
+	}
+
+	// Execute selection.
+	itrs, err := influxql.Select(MustParseSelectStatement(`SELECT (min(value)) FROM cpu WHERE time >= '1970-01-01T00:00:00Z' AND time < '1970-01-02T00:00:00Z' GROUP BY time(10s), host fill(none)`), &ic)
+	if err != nil {
+		t.Fatal(err)
+	} else if a := Iterators(itrs).ReadAll(); !deep.Equal(a, [][]influxql.Point{
+		{&influxql.FloatPoint{Name: "cpu", Tags: ParseTags("host=A"), Time: 0 * Second, Value: 19}},
+		{&influxql.FloatPoint{Name: "cpu", Tags: ParseTags("host=B"), Time: 0 * Second, Value: 10}},
+		{&influxql.FloatPoint{Name: "cpu", Tags: ParseTags("host=A"), Time: 10 * Second, Value: 2}},
+		{&influxql.FloatPoint{Name: "cpu", Tags: ParseTags("host=A"), Time: 30 * Second, Value: 100}},
+	}) {
+		t.Fatalf("unexpected points: %s", spew.Sdump(a))
+	}
+
+	ic.CreateIteratorFn = func(opt influxql.IteratorOptions) (influxql.Iterator, error) {
+		return &FloatIterator{Points: []influxql.FloatPoint{
+			{Name: "cpu", Tags: ParseTags("region=west,host=A"), Time: 0 * Second, Value: 20},
+			{Name: "cpu", Tags: ParseTags("region=west,host=A"), Time: 1 * Second, Value: 19},
+			{Name: "cpu", Tags: ParseTags("region=west,host=B"), Time: 5 * Second, Value: 10},
+			{Name: "cpu", Tags: ParseTags("region=east,host=A"), Time: 9 * Second, Value: 19},
+			{Name: "cpu", Tags: ParseTags("region=east,host=A"), Time: 10 * Second, Value: 2},
+			{Name: "cpu", Tags: ParseTags("region=east,host=A"), Time: 11 * Second, Value: 2},
+			{Name: "cpu", Tags: ParseTags("region=east,host=A"), Time: 12 * Second, Value: 2},
+		}}, nil
+	}
+
+	// Execute selection.
+	itrs, err = influxql.Select(MustParseSelectStatement(`SELECT (distinct(value)) FROM cpu WHERE time >= '1970-01-01T00:00:00Z' AND time < '1970-01-02T00:00:00Z' GROUP BY time(10s), host fill(none)`), &ic)
+	if err != nil {
+		t.Fatal(err)
+	} else if a := Iterators(itrs).ReadAll(); !deep.Equal(a, [][]influxql.Point{
+		{&influxql.FloatPoint{Name: "cpu", Tags: ParseTags("host=A"), Time: 0 * Second, Value: 20}},
+		{&influxql.FloatPoint{Name: "cpu", Tags: ParseTags("host=A"), Time: 1 * Second, Value: 19}},
+		{&influxql.FloatPoint{Name: "cpu", Tags: ParseTags("host=B"), Time: 5 * Second, Value: 10}},
+		{&influxql.FloatPoint{Name: "cpu", Tags: ParseTags("host=A"), Time: 10 * Second, Value: 2}},
+	}) {
+		t.Fatalf("unexpected points: %s", spew.Sdump(a))
+	}
+}

